test/test_wave.py

```diff
@@ -20,4 +20,277 @@
 THE SOFTWARE.
 """
 
-# Tests go here
+import numpy as np
+import numpy.linalg as la
+import pyopencl as cl
+import pyopencl.array as cla  # noqa
+import pyopencl.clmath as clmath
+from pytools.obj_array import join_fields
+import pymbolic as pmbl
+
+from pyopencl.tools import (  # noqa
+        pytest_generate_tests_for_pyopencl
+        as pytest_generate_tests)
+
+import pytest
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+def sym_diff(var):
+    from pymbolic.mapper.differentiator import DifferentiationMapper
+    def func_map(arg_index, func, arg, allowed_nonsmoothness):
+        if func == pmbl.var("sin"):
+            return pmbl.var("cos")(*arg)
+        elif func == pmbl.var("cos"):
+            return -pmbl.var("sin")(*arg)
+        if func == pmbl.var("clsin"):
+            return pmbl.var("clcos")(*arg)
+        elif func == pmbl.var("clcos"):
+            return -pmbl.var("clsin")(*arg)
+        else:
+            raise ValueError("Unrecognized function")
+    return DifferentiationMapper(var, func_map=func_map)
+
+
+def sym_div(vector_func):
+    coord_names = ["x", "y", "z"]
+    div = 0
+    for i in range(len(vector_func)):
+        div = div + sym_diff(pmbl.var(coord_names[i]))(vector_func[i])
+    return div
+
+
+def sym_grad(dim, func):
+    coord_names = ["x", "y", "z"]
+    grad = []
+    for i in range(dim):
+        grad_i = sym_diff(pmbl.var(coord_names[i]))(func)
+        grad.append(grad_i)
+    return grad
+
+
+def sym_eval_with(expr, values):
+    from pymbolic.mapper.evaluator import EvaluationMapper
+    import pymbolic.primitives as prim
+    class mapper(EvaluationMapper):
+        def map_call(self, expr):
+            assert isinstance(expr.function, prim.Variable)
+            if expr.function.name == "sin":
+                par, = expr.parameters
+                return np.sin(self.rec(par))
+            elif expr.function.name == "cos":
+                par, = expr.parameters
+                return np.cos(self.rec(par))
+            if expr.function.name == "clsin":
+                par, = expr.parameters
+                return clmath.sin(self.rec(par))
+            elif expr.function.name == "clcos":
+                par, = expr.parameters
+                return clmath.cos(self.rec(par))
```

Comment:
Rather than distinguishing `cos` and `clcos`, you can just go

```python
par_val = self.rec(par)
from numbers import Number
if isinstance(par_val, Number):
   ...
else:
   ....
```